Sprint 7 solution in file manager

Walkthroughcoding opened the pull request:
Added functionality for task auto-saving to a file

- Implemented the FileBackedTaskManager class for saving and loading tasks.
- Modified task-related methods (addTask, addEpic, addSubtask, updateTask, deleteTask) to call the save() method for automatic file updates.
- Tasks, epics, and subtasks are saved in CSV format.
- Implemented the loadFromFile method to restore tasks from a file on application restart.
- Tested functionality to ensure correct saving and loading of tasks, epics, and subtasks.

UPD 10/02/25
- Made FileBackedTaskManager constructor private and introduced a static `create` method for better encapsulation.
- Added ManagerLoadException to handle file loading errors separately from save exceptions.
- Updated tests in FileBackedTaskManagerTest to use the `create` method for manager initialization.
- Refactored InMemoryHistoryManagerTest to include a @BeforeEach setup method for improved test structure.
- Added additional test cases for edge scenarios in InMemoryHistoryManager (e.g., handling empty history, removing non-existent tasks).

UPD 11/02/25

- Updated FileBackedTaskManager to initialize the manager using loadFromFile(File file)
- Removed redundant create(File file) usage

Walkthroughcoding committed:
1. Add class InMemoryHistoryManager
2. Add interface HistoryManager
3. Add linked list algorithm in InMemoryHistoryManager
4. Add tests for InMemoryHistoryManager

Walkthroughcoding committed:
1. "Removed unused import in TaskManager.java"

Walkthroughcoding committed:
New changes after revert:
1.Add interface HistoryManager
2.Add linked list algorithm in InMemoryHistoryManager
3. Add tests for InMemoryHistoryManager
4. Add HistoryManager getDefaultHistory() in Managers class

Walkthroughcoding committed:
Added functionality for task auto-saving to a file

- Implemented the FileBackedTaskManager class for saving and loading tasks.
- Modified task-related methods (addTask, addEpic, addSubtask, updateTask, deleteTask) to call the save() method for automatic file updates.
- Tasks, epics, and subtasks are saved in CSV format.
- Implemented the loadFromFile method to restore tasks from a file on application restart.
- Tested functionality to ensure correct saving and loading of tasks, epics, and subtasks.

EugeneLenkevich commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,147 @@
+
+            manager.setCurrentId(maxId + 1);
+        } catch (IOException e) {
+            throw new ManagerSaveException("Ошибка при загрузке задач из файла: " + e.getMessage());
```
А еще можно завести аналогичное исключение ManagerLoadException и кидать его тут. Иначе получается что мы кидаем ошибку сохранения при загрузке, не  критично, но не очень логично выглядит :)

EugeneLenkevich commented on test/manager/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,84 @@
+        File tempFile = File.createTempFile("test", ".csv");
+        tempFile.deleteOnExit();
+
+        FileBackedTaskManager manager = new FileBackedTaskManager(tempFile);
```
А давай тут вызовем наш статический метод. И не получится ли спрятать конструктор менеджера задач(с файлом) путем указания ему какого то модификатора доступа например private.
Так мы оставим только один способ создания менеджера, код будет проще.

EugeneLenkevich commented on test/manager/InMemoryHistoryManagerTest.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Test
+    void shouldAddTaskToHistory() {
+        HistoryManager historyManager = Managers.getDefaultHistory();
```
HistoryManager historyManager = Managers.getDefaultHistory();

Можно создать метод, производить в нем вот эту всю инициализацию и не копипастить ее в тестах, повесить на него аннотацию `@Before` и  он станет вызываться перед каждым тестом, будет проще  удобнее :)

EugeneLenkevich reviewed (changes requested):
Привет, Егор! Спасибо за PR.

Отличная работа! :)
Есть совсем небольшие пометки, посмотри комментарии ниже, пожалуйста

Walkthroughcoding committed:
- Made FileBackedTaskManager constructor private and introduced a static `create` method for better encapsulation.
- Added ManagerLoadException to handle file loading errors separately from save exceptions.
- Updated tests in FileBackedTaskManagerTest to use the `create` method for manager initialization.
- Refactored InMemoryHistoryManagerTest to include a @BeforeEach setup method for improved test structure.
- Added additional test cases for edge scenarios in InMemoryHistoryManager (e.g., handling empty history, removing non-existent tasks).

EugeneLenkevich commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -82,7 +86,7 @@ public static FileBackedTaskManager loadFromFile(File file) {
 
```
Подожди, у нас же есть loadFromFile, его тут не получится использовать?

EugeneLenkevich reviewed (changes requested):
Моментик с loadFromFile, не получится ли его использовать для создания менеджера, чтобы не плодить методов?

EugeneLenkevich reviewed (approved):
Пусть у нас будет один способ создания менеджера, спасибо

Walkthroughcoding committed:
- Updated FileBackedTaskManager to initialize the manager using loadFromFile(File file)
- Removed redundant create(File file) usage
